meteor/server/api/cleanup.ts

```diff
@@ -70,134 +82,211 @@ export async function cleanupOldDataInner(actuallyCleanup: boolean = false): Pro
+		removeByQuery(TranslationsBundles, {
+			originBlueprintId: { $nin: blueprintIds },
+		})
 	}
-	// BucketAdLibActions
+
+	// Documents owned by RundownPlaylists:
+	const playlistIds = getAllIdsInCollection(RundownPlaylists, removedRundownPlaylists)
+	const removedRundowns = new Set<RundownId>()
+
 	{
-		ownedByStudioId(BucketAdLibActions)
+		const ownedByRundownPlaylistId = <
+			DBInterface extends { _id: ID; playlistId: RundownPlaylistId },
+			ID extends ProtectedString<any>
+		>(
+			collection: AsyncMongoCollection<DBInterface>
+		): ID[] => {
+			return removeByQuery(collection as AsyncMongoCollection<any>, {
+				playlistId: { $nin: playlistIds },
+			})
+		}
+		ownedByRundownPlaylistId(Rundowns).forEach((id) => removedRundowns.add(id))
+	}
+
+	// Documents owned by Rundowns:
+	const rundownIds = getAllIdsInCollection(Rundowns, removedRundowns)
+	const removedParts = new Set<PartId>()
+	{
+		const ownedByRundownId = <
+			DBInterface extends { _id: ID; rundownId: RundownId },
+			ID extends ProtectedString<any>
+		>(
+			collection: AsyncMongoCollection<DBInterface>
+		): ID[] => {
+			return removeByQuery(collection as AsyncMongoCollection<any>, {
+				rundownId: { $nin: rundownIds },
+			})
+		}
+		ownedByRundownId(AdLibActions)
+		ownedByRundownId(AdLibPieces)
+		ownedByRundownId(IngestDataCache)
+		ownedByRundownId(Parts).forEach((id) => removedParts.add(id))
+		ownedByRundownId(RundownBaselineAdLibActions)
+		ownedByRundownId(RundownBaselineAdLibPieces)
+		ownedByRundownId(RundownBaselineObjs)
+		ownedByRundownId(Segments)
+		// Owned by RundownId Or StudioId:
+		{
+			removeByQuery(ExpectedPlayoutItems, {
+				$or: [
+					{
+						rundownId: { $exists: true, $nin: rundownIds },
+					},
+					{
+						rundownId: { $exists: false },
+						studioId: { $nin: studioIds },
+					},
+				],
+			})
+		}
+		removeByQuery(Pieces, {
+			startRundownId: { $nin: rundownIds },
+		})
+		ownedByRundownId(PieceInstances)
+	}
+	// Documents owned by PeripheralDevices:
+	const removedMediaWorkFlows = new Set<MediaWorkFlowId>()
+	{
+		const deviceIds = getAllIdsInCollection(PeripheralDevices)
+		const ownedByDeviceId = <
+			DBInterface extends { _id: ID; deviceId: PeripheralDeviceId },
+			ID extends ProtectedString<any>
+		>(
+			collection: AsyncMongoCollection<DBInterface>
+		): ID[] => {
+			return removeByQuery(collection as AsyncMongoCollection<any>, {
+				deviceId: { $nin: deviceIds },
+			})
+		}
+		ownedByDeviceId(ExpectedPackageWorkStatuses)
+		ownedByDeviceId(MediaWorkFlows).forEach((id) => removedMediaWorkFlows.add(id))
+		ownedByDeviceId(PackageContainerPackageStatuses)
+		ownedByDeviceId(PackageInfos)
+		ownedByDeviceId(PeripheralDeviceCommands)
 	}
-	// Buckets
+
+	// Documents owned by Parts:
 	{
-		ownedByStudioId(Buckets)
+		const partIds = getAllIdsInCollection(Parts, removedParts)
+		removeByQuery(PartInstances, {
+			$or: [{ rundownId: { $nin: rundownIds } }, { 'part._id': { $nin: partIds } }],
```

Comment:
@Julusian this removal is intended to replace the one you mentioned in the other comment.
However, I'm a tad unsure if this query is safe, let's discuss it a bit?
The original query was
```javascript
{
	reset: true,
	'timings.plannedStoppedPlayback': { $lt: cleanLimitTime },
	'part._id': { $nin: existingPartIds },
}
```
do you think we should apply something similar here?
so
```javascript
{
	reset: true,
	'timings.plannedStoppedPlayback': { $lt: getCurrentTime() - Settings.maximumDataAge },
	$or: [{ rundownId: { $nin: rundownIds } }, { 'part._id': { $nin: partIds } }],
}
```

Reply:
I think this is going to be tricky to combine the two previous behaviours.
The key thing to note is that the `PartInstances` and `PieceInstances` replace the old `AsRunLog` that we used to have.
Which suggests that maybe what we are doing elsewhere is bad, as we remove all the PartInstances when removing a Rundown.

The cleanup was removing any where their Rundown was missing.
The cronjob was removing any that had finished playback and the part they were of was missing. (Intended to catch the TV2DK case of reusing a single rundown for months)

So I think a better representation would be:
```js
{
	$or: [
	{
		// where the parent Rundown is missing
		rundownId: { $nin: rundownIds } 
	},
	{
		// Remove any from long running rundowns where they have long since expired
		reset: true,
		'timings.plannedStoppedPlayback': { $lt: getCurrentTime() - Settings.maximumDataAge },
		'part._id': { $nin: partIds },
	}]
}

```

Similar to your proposal, but better matches what we are doing elsewhere.  
I do feel like yours would be better so that there is a longer living AsRunLog, but that isnt what we are doing elsewhere in Sofie

Reply:
Good points! Let's go with your suggestion